ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/debug/DebuggerStateManager.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Debug session state manager.
+ * 
+ * @author Bartlomiej Laczkowski
+ */
+@Singleton
+public class DebuggerStateManager {
+
+    public static final String LOCAL_STORAGE_DEBUGGER_STATES_KEY_PREFIX = "che-debugger-session-states-";
+
+    private final EventBus     eventBus;
+    private final LocalStorage localStorage;
+    private final DtoFactory   dtoFactory;
+    private final String       storageUniqueKey;
+    private final String       workspaceId;
+
+    @Inject
+    public DebuggerStateManager(AppContext appContext,
+                                WorkspaceServiceClient workspaceServiceClient,
+                                EventBus eventBus,
+                                LocalStorageProvider localStorageProvider,
+                                DtoFactory dtoFactory) {
+        this.eventBus = eventBus;
+        this.localStorage = localStorageProvider.get();
+        this.dtoFactory = dtoFactory;
+        this.workspaceId = appContext.getWorkspaceId();
+        this.storageUniqueKey = LOCAL_STORAGE_DEBUGGER_STATES_KEY_PREFIX + workspaceId;
+        cleanup(workspaceServiceClient);
+        addHandlers();
+    }
+
+    /**
+     * Returns debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debuggerType
+     * @return debugger state
+     */
+    public DebugSessionStateDto getDebuggerState(String debuggerType) {
+        Map<String, DebugSessionStateDto> debugSessionStates = readStateData();
+        return debugSessionStates.get(debuggerType);
+    }
+
+    /**
+     * Sets debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debugSessionState
+     */
+    public void setDebuggerState(DebugSessionStateDto debugSessionState) {
+        Map<String, DebugSessionStateDto> debugSessionStates = readStateData();
+        debugSessionStates.put(debugSessionState.getDebuggerType(), debugSessionState);
+        writeStateData(debugSessionStates);
+    }
+
+    /**
+     * Removes debugger state for given debugger type in scope of current workspace.
+     * 
+     * @param debuggerType
+     */
+    public void removeDebuggerState(String debuggerType) {
+        Map<String, DebugSessionStateDto> debugSessionStates = readStateData();
+        debugSessionStates.remove(debuggerType);
+        writeStateData(debugSessionStates);
+    }
+
+    private void addHandlers() {
+        eventBus.addHandler(WorkspaceStoppedEvent.TYPE, new WorkspaceStoppedEvent.Handler() {
+            @Override
+            public void onWorkspaceStopped(WorkspaceStoppedEvent event) {
+                if (workspaceId.equals(event.getWorkspace().getId())) {
+                    // Wipe out storage data key on workspace stop
+                    writeStateData(null);
+                }
+            }
+        });
+    }
+
+    private Map<String, DebugSessionStateDto> readStateData() {
+        if (localStorage == null) {
+            return new HashMap<>();

```

Comment:
Yes we need, for the very first usage some data will be added to this map and Collections.emptyMap() won't allow it as it is immutable.